[18.0][MIG] base_technical_user: Migration to 18.0

Kimkhoi3010 opened the pull request:
Note:

- Missing commits are ported: [[17.0][FW] base_technical_user: port from 14.0](https://github.com/OCA/server-tools/pull/3068)

florian-dacosta committed:
Add onchange_helper module

alexis-via committed:
[MIG] module onchange_helper to v10.0

alexis-via committed:
[FIX] avoid a crash when value is False for a M2O field

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[10.0] add new module to manage technical user on company

[CHG] add name to wml tag to make override easier

[FIX] remove domain on active field

oca-transbot committed:
OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

astirpe committed:
[11.0][MIG] onchange_helper + add tests

hugosantosred committed:
[MIG] base_technical_user to V11

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-travis committed:
[UPD] Update onchange_helper.pot

osoul-dev committed:
Added translation using Weblate (Arabic)

simahawk committed:
Global pylint cleanup

sbejaoui committed:
[12.0][MIG] - onchange_helper migration to 12.0

sebastienbeau committed:
[FWP] - Forward porting changes from #1253

sbejaoui committed:
[IMP] - Adapt unit tests

sbejaoui committed:
[IMP] - read may be harmful to performance

sbejaoui committed:
[IMP] - get default values for non record onchange call

other improvement:

set onchange_specs to all fields as _onchange_spec()
retrun onchange fields for default view
return field value if it's set in onchange_fields (usuful to get default value for a field)

sbejaoui committed:
[UPD] - Update documentation

bguillot committed:
[FIX] don't remove computed fields with inverse method of the onchange values

OCA-git-bot committed:
[ADD] icon.png

florian-dacosta committed:
Migrate base_technical_user to version 12

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[ADD] icon.png

oca-travis committed:
[UPD] Update onchange_helper.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-12.0/server-tools-12.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-12-0/server-tools-12-0-base_technical_user/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-12.0/server-tools-12.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-12-0/server-tools-12-0-onchange_helper/

OCA-git-bot committed:
[UPD] README.rst

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 100.0% (5 of 5 strings)

Translation: server-tools-12.0/server-tools-12.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-12-0/server-tools-12-0-base_technical_user/pt/

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (5 of 5 strings)

Translation: server-tools-12.0/server-tools-12.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-12-0/server-tools-12-0-base_technical_user/zh_CN/

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (1 of 1 strings)

Translation: server-tools-12.0/server-tools-12.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-12-0/server-tools-12-0-onchange_helper/zh_CN/

Commit:
[IMP] onchange_helper: black, isort

Commit:
[MIG] onchange_helper: Migration to 13.0

oca-travis committed:
[UPD] Update onchange_helper.pot

OCA-git-bot committed:
[UPD] README.rst

sebastienbeau committed:
[IMP] add helper for getting a self env with the tech user

HviorForgeFlow committed:
[IMP] base_technical_user: black, isort, prettier

HviorForgeFlow committed:
[MIG] base_technical_user: Migration to 13.0

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-13.0/server-tools-13.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-13-0/server-tools-13-0-base_technical_user/

kevinkhao committed:
[14.0][MIG] onchange_helper: Migration to 14.0

Commit:
[IMP] base_technical_user: black, isort, prettier

Commit:
[MIG] base_technical_user: Migration to 14.0

oca-travis committed:
[UPD] Update onchange_helper.pot

OCA-git-bot committed:
[UPD] README.rst

pedrocs-exo committed:
Translated using Weblate (Portuguese)

Currently translated at 85.7% (6 of 7 strings)

Translation: server-tools-14.0/server-tools-14.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-14-0/server-tools-14-0-base_technical_user/pt/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-14.0/server-tools-14.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-14-0/server-tools-14-0-base_technical_user/

Yvesldff committed:
Translated using Weblate (French)

Currently translated at 54.5% (6 of 11 strings)

Translation: server-tools-14.0/server-tools-14.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-14-0/server-tools-14-0-base_technical_user/fr/

rousseldenis committed:
[14.0] base_technical_user: Allow multi company context

First, the company should be taken through self.env.company as user.company_id
is now the default company for the user (and not the current one).

Secondly, the actual code does not allow to do recordset.with_company(B).sudo_tech()....
as allowed_company_ids in context can be filled in with current company
(and not the one of the tech user - which is unique)

rousseldenis committed:
[14.0][IMP] base_technical_user: Multi company context, add tests

koenloodts committed:
[MIG] onchange_helper: Migration to 15.0

rousseldenis committed:
[14.0][IMP] onchange_helper: Call onchange only if fields are concerned by

As the onchange method can be a little bulky, we don't call it if fields passed as
arguments are not in _onchange_methods

Commit:
[UPD] Update onchange_helper.pot

OCA-git-bot committed:
[UPD] README.rst

simahawk committed:
Replace base_jsonify w/ jsonifier

Rationale:

1. the name reflects better what it does
   as this is not _just_ a base module

2. solves publication issue on odoo apps store
   due to an old module registered w/ the same name
   that even if un-published blocks publishing the module
   for all versions.
   This in turn, blocks the publication of ALL modules
   that depend on base_jsonify.

simahawk committed:
Add jsonifier (skeleton)

OCA-git-bot committed:
[ADD] icon.png

Commit:
[UPD] Update jsonifier.pot

OCA-git-bot committed:
[UPD] README.rst

simahawk committed:
jsonifier: fix date handling

We should always keep UTC and let the consumer deal w/ it.
This change is backward compatible but allows to turn off
the old behavior w/ the ctx flag `jsonifier__date_user_tz=False`.

OCA-git-bot committed:
jsonifier 14.0.1.0.1

simahawk committed:
jsonifier: reduce complexity of _jsonify_record

simahawk committed:
jsonifier: allow json_key override by resolver

Quite handy to take full control of the final result.

florian-dacosta committed:
[FIX] Keep computed editable field value when playing onchanges

OCA-git-bot committed:
onchange_helper 15.0.1.0.1

ibuioli committed:
Added translation using Weblate (Spanish (Argentina))

ibuioli committed:
Translated using Weblate (Spanish (Argentina))

Currently translated at 100.0% (1 of 1 strings)

Translation: server-tools-15.0/server-tools-15.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-15-0/server-tools-15-0-onchange_helper/es_AR/

HviorForgeFlow committed:
[MIG] base_technical_user: Migration to 15.0

sebastienbeau committed:
[IMP] jsonifier: black, isort, prettier

sebastienbeau committed:
jsonifier: migration to v16

kevinkhao committed:
[MIG] onchange_helper: migration to 16.0

danielduqma committed:
[16.0][MIG] base_technical_user

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-16.0/server-tools-16.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-base_technical_user/

Commit:
[UPD] Update onchange_helper.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-16.0/server-tools-16.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-onchange_helper/

sebastienbeau committed:
jsonifier: remove deprecated logic for getting the date in user tz

Commit:
[UPD] Update jsonifier.pot

OCA-git-bot committed:
[UPD] README.rst

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-16.0/server-tools-16.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-jsonifier/

xavier-bouquiaux committed:
[UPD] onchange_helper - AGPL to LGPL

see issue #2542

Commit:
[UPD] Update jsonifier.pot

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-16.0/server-tools-16.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-base_technical_user/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-16.0/server-tools-16.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-jsonifier/

kikopeiro committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: server-tools-16.0/server-tools-16.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-base_technical_user/es/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (1 of 1 strings)

Translation: server-tools-16.0/server-tools-16.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-onchange_helper/es/

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

gfcapalbo committed:
[ADD] use jsonifier features to retund postprocessed values, containing
field string as well

gfcapalbo committed:
Better Version, not using function features , leaves parser syntax
intact.

gfcapalbo committed:
[FIX] Review fixes

gfcapalbo committed:
[FIX] small comment and style fix

Ivorra78 committed:
Added translation using Weblate (Spanish)

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (39 of 39 strings)

Translation: server-tools-16.0/server-tools-16.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-jsonifier/es/

acsonefho committed:
[FIX] base_technical_user - Error message when no technical user.
When we use use_tech(raise_if_missing=True) and the technical user (user_tech_id) is not defined on the current company, the error message who said that field is empty should be related to the active company and not related to the default user's company.

nguyenminhchien committed:
[IMP] jsonifier: pre-commit auto fixes

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (1 of 1 strings)

Translation: server-tools-16.0/server-tools-16.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-onchange_helper/it/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: server-tools-16.0/server-tools-16.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-16-0/server-tools-16-0-base_technical_user/it/

nguyenminhchien committed:
[MIG] jsonifier: Migration to 17.0

john-herholz-dt committed:
[IMP] base_technical_user: pre_commit auto fixes

john-herholz-dt committed:
[MIG] base_technical_user: Migration to 17.0

john-herholz-dt committed:
[IMP] onchange_helper: pre-commit auto fixes

john-herholz-dt committed:
[MIG] onchange_helper: Migration to 17.0

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[UPD] Update jsonifier.pot

OCA-git-bot committed:
[BOT] post-merge updates

Commit:
[UPD] Update onchange_helper.pot

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-17.0/server-tools-17.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-17-0/server-tools-17-0-base_technical_user/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-17.0/server-tools-17.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-17-0/server-tools-17-0-jsonifier/

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (37 of 37 strings)

Translation: server-tools-17.0/server-tools-17.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-17-0/server-tools-17-0-jsonifier/it/

Rad0van committed:
[FIX] Add missing depend

Without explicit web dependency all tests fail like this:

File "/home/odoo/src/odoo/odoo/models.py", line 6905, in onchange
    raise NotImplementedError("onchange() is implemented in module 'web'")
NotImplementedError: onchange() is implemented in module 'web'

Rad0van committed:
Update __manifest__.py

OCA-git-bot committed:
[BOT] post-merge updates

xtanuiha committed:
Translated using Weblate (Chinese (Simplified) (zh_CN))

Currently translated at 100.0% (7 of 7 strings)

Translation: server-tools-17.0/server-tools-17.0-base_technical_user
Translate-URL: https://translation.odoo-community.org/projects/server-tools-17-0/server-tools-17-0-base_technical_user/zh_CN/

xtanuiha committed:
Translated using Weblate (Chinese (Simplified) (zh_CN))

Currently translated at 100.0% (37 of 37 strings)

Translation: server-tools-17.0/server-tools-17.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-17-0/server-tools-17-0-jsonifier/zh_CN/

Kimkhoi3010 committed:
[MIG] base_technical_user: Migration to 18.0

thienvh332 committed:
[IMP] onchange_helper: apply pre-commit auto fixes

thienvh332 committed:
[MIG] onchange_helper: Migration to 18.0

thienvh332 committed:
[IMP] jsonifier: apply pre-commit auto fixes

thienvh332 committed:
[MIG] jsonifier: Migration to 18.0

OCA-git-bot committed:
Initial commit

henrybackman commented on base_technical_user/models/models.py:
```diff
@@ -0,0 +1,36 @@
+class Base(models.AbstractModel):
+    _inherit = "base"
+
+    def sudo_tech(self, raise_if_missing=False):
+        self_sudoer = self
+        tech_user = self.env.company.user_tech_id
+        if tech_user:
+            self_sudoer = self.with_user(tech_user.id)
+            # We restrict the allowed companies to the one of the tech user
+            allowed_company_ids = self.env.context.get("allowed_company_ids")
+            # TODO: Is any(...) part necessary as we can consider company should always be
+            # the one of the tech_user ?
+            if allowed_company_ids and any(
+                company
+                for company in allowed_company_ids
+                if company != tech_user.company_id.id
+            ):
+                self_sudoer = self_sudoer.with_context(
+                    allowed_company_ids=self.env.company.ids
+                )
+        elif raise_if_missing:
+            raise UserError(
+                _("The technical user is missing in the company {}").format(
+                    self.env.company.name
+                )
+            )
+        return self_sudoer
```
Only version 14 has this modified version of this, 15, 16 and 17 are like this:
```
    def sudo_tech(self, raise_if_missing=False):
        self_sudoer = self
        tech_user = self.env.user.company_id.user_tech_id
        if tech_user:
            self_sudoer = self.with_user(tech_user.id)
        elif raise_if_missing:
            raise UserError(
                _("The technical user is missing in the company {}").format(
                    self.env.user.company_id.name
                )
            )
        return self_sudoer
```

Kimkhoi3010 replied:
Thank you for the reminder.

henrybackman commented on base_technical_user/readme/CREADITS.md:
Misspelled file name, should be `CREDITS.md`

Kimkhoi3010 replied:
That was my big mistake. Thank you for the reminder.

simahawk commented:
/ocabot migration base_technical_user

OCA-git-bot committed:
Merge PR #3053 into 18.0

Signed-off-by simahawk

Commit:
[UPD] Update jsonifier.pot

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
Merge PR #3052 into 18.0

Signed-off-by simahawk

Commit:
[UPD] Update onchange_helper.pot

simahawk commented:
@Kimkhoi3010 pls check the CI and fixups

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-18.0/server-tools-18.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-18-0/server-tools-18-0-jsonifier/

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: server-tools-18.0/server-tools-18.0-onchange_helper
Translate-URL: https://translation.odoo-community.org/projects/server-tools-18-0/server-tools-18-0-onchange_helper/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (37 of 37 strings)

Translation: server-tools-18.0/server-tools-18.0-jsonifier
Translate-URL: https://translation.odoo-community.org/projects/server-tools-18-0/server-tools-18-0-jsonifier/it/

simahawk commented:
/ocabot merge nobump

Kimkhoi3010 commented:
@simahawk, @henrybackman 
I created a new PR [OCA#3074](https://github.com/OCA/server-tools/pull/3074) to OCA/server-tools.
(This PR was closed auto after rebased)
